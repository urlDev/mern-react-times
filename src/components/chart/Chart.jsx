import React from 'react';
import * as d3 from 'd3';
import { useSelector } from 'react-redux';

import { ChartContainer } from './Chart.styles';

const Chart = () => {
  const { chartData } = useSelector((chart) => chart.chart);
  const d3Ref = React.useRef(null);

  React.useEffect(() => {
    const svg = d3.select(d3Ref.current);

    const width = 550;
    const height = 345;
    const padding = 50;

    svg.attr('width', width).attr('height', height);

    const dates = chartData.map((value) => new Date(value.date));

    // xScale
    const xAxisScale = d3
      .scaleTime()
      .domain([d3.min(dates), d3.max(dates)])
      .range([padding, width - padding]);

    // yScale
    const yAxisScale = d3
      .scaleLinear()
      .domain([
        d3.min(chartData, (data) => data.close),
        d3.max(chartData, (data) => data.close),
      ])
      .range([height - padding, padding]);

    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(5)
      .tickFormat((date) => d3.timeFormat('%b %d, %I:%M')(date));

    // draw x axes
    svg
      .append('g')
      .call(xAxes)
      .attr('transform', `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    svg.append('g').call(yAxes).attr('transform', `translate(${padding},0)`);

    // create line
    const line = d3
      .line()
      .x((d) => xAxisScale(new Date(d.date)))
      .y((d) => yAxisScale(d.close));

    // draw historical price path
    const path = svg
      .append('path')
      .data([chartData])
      .style('fill', ' none')
      .attr('stroke', 'black')
      .attr('stroke-width', '1.5')
      .attr('d', line);

    // get the length of the path for line transition
    const pathLength = path.node().getTotalLength();

    // transition for the line
    const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

    // setting pathsLength for strokes attrs and using transition
    // I made it minus pathLength because it was drawing it backwards
    path
      .attr('stroke-dashoffset', -pathLength)
      .attr('stroke-dasharray', pathLength)
      .transition(transitionPath)
      .attr('stroke-dashoffset', 0);

    path.interrupt('stroke-dashoffset');
    path.interrupt('stroke-dasharray');

    return () => {
      // clean the canvas for the next chart
      svg.selectAll('*').remove();
    };
  }, [chartData]);

  return (
    <ChartContainer>
      <svg ref={d3Ref}></svg>
    </ChartContainer>
  );
};

export default Chart;
