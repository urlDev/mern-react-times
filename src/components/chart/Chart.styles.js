import styled from "styled-components";

export const ChartContainer = styled.div`
  padding-right: 40px;
  position: relative;
  &:after {
    content: "";
    position: absolute;
    top: 5%;
    right: 0;
    width: 1px;
    height: 90%;
    background: lightgray;
  }

  @media (max-width: 1400px) {
    display: flex;
    justify-content: center;
    padding: 0;
    &:after {
      display: none;
    }
  }
`;
