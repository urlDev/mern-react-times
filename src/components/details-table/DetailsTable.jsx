import { useSelector } from 'react-redux';

import { DetailsTableContainer } from './DetailsTable.styles';

const DetailsTable = () => {
  const { marketDetail, rating } = useSelector((chart) => chart.chart);

  const {
    symbol,
    price,
    marketCap,
    yearHigh,
    yearLow,
    volume,
    changesPercentage,
    priceAvg50,
    priceAvg200,
  } = marketDetail;

  const convertNumber = (labelValue) => {
    // Nine Zeroes for Billions
    return Math.abs(Number(labelValue)) >= 1.0e9
      ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + 'B'
      : // Six Zeroes for Millions
      Math.abs(Number(labelValue)) >= 1.0e6
      ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + 'M'
      : // Three Zeroes for Thousands
      Math.abs(Number(labelValue)) >= 1.0e3
      ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + 'K'
      : Math.abs(Number(labelValue));
  };

  return (
    <DetailsTableContainer>
      {marketDetail.symbol ? (
        <table>
          <tbody>
            <tr>
              <td>Symbol</td>
              <td>{symbol.split('^').join('')}</td>
            </tr>
            <tr>
              <td>Price</td>
              <td>$ {price}</td>
            </tr>
            <tr>
              <td>Percentage Change</td>
              <td>%{changesPercentage}</td>
            </tr>
            <tr>
              <td>Price Average 50</td>
              <td>$ {priceAvg50.toFixed(3)}</td>
            </tr>
            <tr>
              <td>Price Average 200</td>
              <td>$ {priceAvg200.toFixed(3)}</td>
            </tr>
            <tr>
              <td>Market Cap</td>
              <td>$ {convertNumber(marketCap)}</td>
            </tr>
            <tr>
              <td>Year High</td>
              <td>$ {yearHigh}</td>
            </tr>
            <tr>
              <td>Year Low</td>
              <td>$ {yearLow}</td>
            </tr>
            <tr>
              <td>Volume</td>
              <td>{convertNumber(volume)}</td>
            </tr>
          </tbody>
        </table>
      ) : null}
      {rating.length > 0 && (
        <table>
          <tbody>
            <tr>
              <td>Rating</td>
              <td>{rating[0].ratingRecommendation}</td>
            </tr>
            <tr>
              <td>DCF Recommendation</td>
              <td>{rating[0].ratingDetailsDCFRecommendation}</td>
            </tr>
            <tr>
              <td>DE Recommendation</td>
              <td>{rating[0].ratingDetailsDERecommendation}</td>
            </tr>
            <tr>
              <td>PB Recommendation</td>
              <td>{rating[0].ratingDetailsPBRecommendation}</td>
            </tr>
            <tr>
              <td>PE Recommendation</td>
              <td>{rating[0].ratingDetailsPERecommendation}</td>
            </tr>
            <tr>
              <td>ROA Recommendation</td>
              <td>{rating[0].ratingDetailsROARecommendation}</td>
            </tr>
            <tr>
              <td>ROE Recommendation</td>
              <td>{rating[0].ratingDetailsROERecommendation}</td>
            </tr>
          </tbody>
        </table>
      )}
    </DetailsTableContainer>
  );
};

export default DetailsTable;
