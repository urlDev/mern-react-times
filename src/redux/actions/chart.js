import axios from "axios";

export const FETCH_FOREX_SUCCESS = "FETCH_FOREX_SUCCESS";
export const FETCH_CHART_ERROR = "FETCH_CHART_ERROR";
export const SET_LOADING = "SET_LOADING";
export const SET_LOADING_TRUE = "SET_LOADING_TRUE";
export const CHANGE_MARKET_TYPE = "CHANGE_MARKET_TYPE";
export const CHANGE_MARKET_NAME = "CHANGE_MARKET_NAME";
export const SET_MARKET_DETAIL = "SET_MARKET_DETAIL";
export const SET_CHART_DATA = "SET_CHART_DATA";
export const CLEAN_CHART_DATA = "CLEAN_CHART_DATA";
export const SET_CHART_TIME_FRAME = "SET_CHART_TIME_FRAME";
export const SET_RATING = "SET_RATING";
export const SET_HOME_CHART_DATA = "SET_HOME_CHART_DATA";
export const CLEAN_STATE = "CLEAN_STATE";
export const GET_SEARCH_RESULTS = "GET_SEARCH_RESULTS";
export const SET_SEARCH_MARKET_DETAIL = "SET_SEARCH_MARKET_DETAIL";
export const CLEAR_SEARCH_RESULTS = "CLEAR_SEARCH_RESULTS";
export const OPEN_SEARCH_MODAL = "OPEN_SEARCH_MODAL";
export const CLOSE_SEARCH_MODAL = "CLOSE_SEARCH_NODAL";

export const closeSearchModal = () => ({
  type: CLOSE_SEARCH_MODAL,
});

export const openSearchModal = () => ({
  type: OPEN_SEARCH_MODAL,
});

export const clearSearchResults = () => ({
  type: CLEAR_SEARCH_RESULTS,
});

export const setSearchMarketDetail = (details) => ({
  type: SET_SEARCH_MARKET_DETAIL,
  payload: details,
});

export const getSearchResults = (result) => ({
  type: GET_SEARCH_RESULTS,
  payload: result,
});

export const setLoading = () => ({
  type: SET_LOADING,
});

export const setLoadingTrue = () => ({
  type: SET_LOADING_TRUE,
});

export const cleanState = () => ({
  type: CLEAN_STATE,
});

export const setHomeChartData = (chart) => ({
  type: SET_HOME_CHART_DATA,
  payload: [chart],
});

export const setRating = (rating) => ({
  type: SET_RATING,
  payload: rating,
});

export const setChartTimeFrame = (time) => ({
  type: SET_CHART_TIME_FRAME,
  payload: time,
});

export const setChartData = (data) => ({
  type: SET_CHART_DATA,
  payload: data,
});

export const cleanChartData = () => ({
  type: CLEAN_CHART_DATA,
});

export const setMarketDetail = (market) => ({
  type: SET_MARKET_DETAIL,
  payload: market,
});

export const changeMarketType = (market) => ({
  type: CHANGE_MARKET_TYPE,
  payload: market,
});

export const changeMarketName = (market) => ({
  type: CHANGE_MARKET_NAME,
  payload: market,
});

export const fetchForexSuccess = (chart) => ({
  type: FETCH_FOREX_SUCCESS,
  payload: chart,
});

export const fetchChartError = (error) => ({
  type: FETCH_CHART_ERROR,
  payload: error,
});

export const fetchForex = (market) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/quote/${market}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    // FinancialModel API doesn't send error if it's about limit reach.
    // In that sense, they send an error message with 200 status. Weird.
    // Thats why I am catching that error here and dispatching it.
    if (data.length) {
      return data.length === 1
        ? [dispatch(setSearchMarketDetail(data))]
        : dispatch(fetchForexSuccess(data));
    } else {
      return dispatch(fetchChartError(data));
    }
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchChartData = (symbol, timeFrame) => async (dispatch) => {
  dispatch(cleanChartData());
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/historical-chart/${timeFrame}/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setChartData(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchRating = (symbol) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/rating/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setRating(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

/*
This deserves to be mentioned.
Few months ago, when I was making the first version of this app, I used array.map to fetch charts.
Because of the async nation and using array map method, response was coming irregular and I had to sort them.
Now, with using for of loop, I dont need to write any extra code because loop waits for one to finish to move on.
Plus, its so clean!
Yet another homerun for me! 😁😎
*/
export const fetchHomeChart = (symbols) => async (dispatch) => {
  const symbolsArray = symbols.split(",").sort();
  let charts = [];
  try {
    // dispatch(setLoadingTrue());
    dispatch(cleanState());
    for (let symbol of symbolsArray) {
      const response = await axios.get(
        `https://financialmodelingprep.com/api/v3/historical-chart/1hour/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
      );
      const data = response.data;
      charts.push(data);
    }
    return charts.length
      ? dispatch(setHomeChartData(charts))
      : dispatch(fetchChartError(charts));
    // dispatch(setLoading());
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchSearch = (input) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/search?query=${input}&limit=10&apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = response.data;
    return dispatch(getSearchResults(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};
