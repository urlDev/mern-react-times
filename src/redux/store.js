import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));