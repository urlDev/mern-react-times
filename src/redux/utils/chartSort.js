const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;