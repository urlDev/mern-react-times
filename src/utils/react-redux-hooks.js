// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
