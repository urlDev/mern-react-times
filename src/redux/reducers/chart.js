import {
    FETCH_FOREX_SUCCESS,
    FETCH_CHART_ERROR,
    CHANGE_MARKET_TYPE,
    SET_MARKET_DETAIL,
    SET_CHART_DATA,
    SET_CHART_TIME_FRAME,
    SET_RATING,
    SET_HOME_CHART_DATA,
} from '../actions/chart';

import chartSort from '../utils/chartSort';

const initialState = {
    loading: true,
    forex: [],
    error: null,
    // Indexes are first to load.
    marketType: '%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE',
    marketDetail: {},
    chartData: [],
    chartTimeFrame: '1hour',
    rating: [],
    homeChartData: [],
};

const chartReducer = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_FOREX_SUCCESS:
            return {
                ...state,
                forex: action.payload,
                loading: false,
            };
        case FETCH_CHART_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case CHANGE_MARKET_TYPE:
            return {
                ...state,
                marketType: action.payload,
            };
        case SET_MARKET_DETAIL:
            return {
                ...state,
                marketDetail: action.payload.data,
            };
        case SET_CHART_DATA:
            return {
                ...state,
                chartData: action.payload,
                loading: false,
            };
        case SET_CHART_TIME_FRAME:
            return {
                ...state,
                chartTimeFrame: action.payload,
            };
        case SET_RATING:
            return {
                ...state,
                rating: action.payload,
            };
        case SET_HOME_CHART_DATA:
            return {
                ...state,
                homeChartData: [...state.homeChartData, action.payload],
                loading: false,
            };
        default:
            return state;
    }
};

export default chartReducer;