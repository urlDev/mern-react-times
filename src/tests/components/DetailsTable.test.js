import { shallow } from "enzyme";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import { marketDetail, rating } from "tests/fixtures/chart";

import DetailsTable from "components/details-table/DetailsTable";

describe("Testing DetailsTable component", () => {
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore({
      marketDetail,
      rating,
    });

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation(() => store.getState());

    wrapper = shallow(<DetailsTable store={store} />);
  });

  test("Should render the component with marketDetail and rating", () => {
    expect(wrapper).toHaveLength(1);
    expect(wrapper.find("td").first().text()).toEqual("Symbol");
    expect(wrapper.find("td").last().text()).toEqual(
      rating[0].ratingDetailsROERecommendation
    );
  });
  test("Should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Should not show rating if it doesnt exist", () => {
    store = mockStore({
      marketDetail,
      rating: [],
    });

    wrapper = shallow(<DetailsTable store={store} />);

    expect(wrapper).toHaveLength(1);
    expect(wrapper.find("td").last().text()).not.toEqual(
      rating[0].ratingDetailsROERecommendation
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("Should not show marketDetails if it doesnt exist", () => {
    store = mockStore({
      marketDetail: [],
      rating,
    });

    wrapper = shallow(<DetailsTable store={store} />);

    expect(wrapper).toHaveLength(1);
    expect(wrapper.find("td").first().text()).not.toEqual("Symbol");
    expect(wrapper).toMatchSnapshot();
  });
});
