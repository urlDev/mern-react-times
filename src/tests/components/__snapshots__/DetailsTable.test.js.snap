// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DetailsTable component Should match the snapshot 1`] = `
<styled.div>
  <table>
    <tbody>
      <tr>
        <td>
          Symbol
        </td>
        <td>
          TSLA
        </td>
      </tr>
      <tr>
        <td>
          Price
        </td>
        <td>
          $ 
          665.31
        </td>
      </tr>
      <tr>
        <td>
          Percentage Change
        </td>
        <td>
          %
          -4.27
        </td>
      </tr>
      <tr>
        <td>
          Price Average 50
        </td>
        <td>
          $ 
          530.535
        </td>
      </tr>
      <tr>
        <td>
          Price Average 200
        </td>
        <td>
          $ 
          387.145
        </td>
      </tr>
      <tr>
        <td>
          Market Cap
        </td>
        <td>
          $ 
          630.65B
        </td>
      </tr>
      <tr>
        <td>
          Year High
        </td>
        <td>
          $ 
          695
        </td>
      </tr>
      <tr>
        <td>
          Year Low
        </td>
        <td>
          $ 
          70.102
        </td>
      </tr>
      <tr>
        <td>
          Volume
        </td>
        <td>
          38.94M
        </td>
      </tr>
    </tbody>
  </table>
  <table>
    <tbody>
      <tr>
        <td>
          Rating
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          DCF Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          DE Recommendation
        </td>
        <td>
          Buy
        </td>
      </tr>
      <tr>
        <td>
          PB Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          PE Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          ROA Recommendation
        </td>
        <td>
          Neutral
        </td>
      </tr>
      <tr>
        <td>
          ROE Recommendation
        </td>
        <td>
          Neutral
        </td>
      </tr>
    </tbody>
  </table>
</styled.div>
`;

exports[`Testing DetailsTable component Should not show marketDetails if it doesnt exist 1`] = `
<styled.div>
  <table>
    <tbody>
      <tr>
        <td>
          Rating
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          DCF Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          DE Recommendation
        </td>
        <td>
          Buy
        </td>
      </tr>
      <tr>
        <td>
          PB Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          PE Recommendation
        </td>
        <td>
          Strong Buy
        </td>
      </tr>
      <tr>
        <td>
          ROA Recommendation
        </td>
        <td>
          Neutral
        </td>
      </tr>
      <tr>
        <td>
          ROE Recommendation
        </td>
        <td>
          Neutral
        </td>
      </tr>
    </tbody>
  </table>
</styled.div>
`;

exports[`Testing DetailsTable component Should not show rating if it doesnt exist 1`] = `
<styled.div>
  <table>
    <tbody>
      <tr>
        <td>
          Symbol
        </td>
        <td>
          TSLA
        </td>
      </tr>
      <tr>
        <td>
          Price
        </td>
        <td>
          $ 
          665.31
        </td>
      </tr>
      <tr>
        <td>
          Percentage Change
        </td>
        <td>
          %
          -4.27
        </td>
      </tr>
      <tr>
        <td>
          Price Average 50
        </td>
        <td>
          $ 
          530.535
        </td>
      </tr>
      <tr>
        <td>
          Price Average 200
        </td>
        <td>
          $ 
          387.145
        </td>
      </tr>
      <tr>
        <td>
          Market Cap
        </td>
        <td>
          $ 
          630.65B
        </td>
      </tr>
      <tr>
        <td>
          Year High
        </td>
        <td>
          $ 
          695
        </td>
      </tr>
      <tr>
        <td>
          Year Low
        </td>
        <td>
          $ 
          70.102
        </td>
      </tr>
      <tr>
        <td>
          Volume
        </td>
        <td>
          38.94M
        </td>
      </tr>
    </tbody>
  </table>
</styled.div>
`;
